### PR TITLE
[PyTorch][MoE] Kernels fusions for the MoE router

### PR DESCRIPTION
**Pull request by Autumn1998**
# Description

Provide function used in the router fusion and corresponding unit-test.

1. fuse the topk+softmax/sigmoid
2. fuse the score function used in the aux loss
3. fuse the aux loss computation

All 3 parts include the forward and backward.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/pytorch/test_fused_router.py
Comment (timmoon10) [inline tests/pytorch/test_fused_router.py:1]: We should also add this test to this QA script: https://github.com/NVIDIA/TransformerEngine/blob/main/qa/L0_pytorch_unittest/test.sh

File: tests/pytorch/test_fused_router.py [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+        topk=topk,
+        use_pre_softmax=use_pre_softmax,
+        num_groups=num_groups,
+        group_topk=group_topk,
+        scaling_factor=scaling_factor,
+        score_function=score_function,
+        expert_bias=expert_bias_clone,
+    )
+
+    assert torch.allclose(probs, probs_fused, atol=atol, rtol=rtol)
```
Comment (timmoon10) [inline tests/pytorch/test_fused_router.py]: It would be nicer to replace `torch.allclose` with [`torch.testing.assert_close`](https://docs.pytorch.org/docs/stable/testing.html#torch.testing.assert_close). It has a helpful error message and it automatically chooses the tols based on the dtype.
```suggestion
    torch.testing.assert_close(probs, probs_fused)
```

File: tests/pytorch/test_fused_router.py [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+
+
+@pytest.mark.parametrize("dtype", [torch.float32])
+@pytest.mark.parametrize("num_tokens", [2048, 7168, 32111])
+@pytest.mark.parametrize("num_experts", [128, 32])
+@pytest.mark.parametrize("topk", [4, 8])
+@pytest.mark.parametrize("group_topk", [None, 4])
+@pytest.mark.parametrize("scaling_factor", [None, 1.2])
+@pytest.mark.parametrize("enable_bias", [True, False])
+def test_topk_sigmoid(
```
Comment (timmoon10) [inline tests/pytorch/test_fused_router.py:219]: How long does this test suite take to run? The number of test cases grows very quickly if you have one test with many parameters (O(2^n) cases), so it may be better to split it up into multiple tests with only a few parameters (O(2*n) cases).

File: transformer_engine/common/include/transformer_engine/fused_router.h [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+ *  \param[in]     num_tokens      Number of tokens.
+ *  \param[in]     num_experts     Number of experts.
+ *  \param[in]     topk            Topk value.
+ *  \param[in]     score_function  Score function, 0: sigmoid, 1: softmax.
+ *  \param[out]    scores          Output tensor for scores.
+ *  \param[in]     routing_map     Routing map.
+ *  \param[in]     intermediate_output  Intermediate output from the forward pass. (Softmax/sigmoid output)
+ *  \param[in]     stream          CUDA stream used for the operation.
+ */
+void nvte_fused_scores_for_aux_loss_forward(const NVTETensor logits, int num_tokens,
```
Comment (timmoon10) [inline transformer_engine/common/include/transformer_engine/fused_router.h]: What is the naming convention for this loss function in other MoE implementations? It feels like `aux_loss` is too general and it might get confusing if some other multi-objective training method becomes popular in the future. Maybe something like `moe_aux_loss` would be more specific.

File: tests/pytorch/test_fused_router.py (hunk 1 above)
Comment (Autumn1998) [reply tests/pytorch/test_fused_router.py]: @timmoon10 updated

File: tests/pytorch/test_fused_router.py (hunk 2 above)
Comment (Autumn1998) [reply tests/pytorch/test_fused_router.py:219]: The kernels are small and will not consume long time . All 237 test cases only need several seconds(<5s).

File: transformer_engine/common/include/transformer_engine/fused_router.h (hunk 3 above)
Comment (Autumn1998) [reply transformer_engine/common/include/transformer_engine/fused_router.h]: I have update all aux_loss to moe_aux_loss.

File: tests/pytorch/test_fused_router.py (hunk 2 above)
Comment (Autumn1998) [reply tests/pytorch/test_fused_router.py:219]: @timmoon10

File: tests/pytorch/test_fused_router.py
Comment (timmoon10) [reply tests/pytorch/test_fused_router.py:1]: Seems like the test was removed from the QA script.

File: transformer_engine/common/include/transformer_engine/fused_router.h [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ *  \param[in]     scaling_factor  Scaling factor.
+ *  \param[in]     score_function  Score function, 0: sigmoid, 1: softmax.
+ *  \param[in]     expert_bias     Expert bias. (Only used at the sigmoid case)
+ *  \param[out]    probs           Output tensor for probabilities.
+ *  \param[out]    routing_map     Output tensor for routing map.
+ *  \param[out]    intermediate_output  Output tensor for intermediate output. (Softmax/sigmoid output)
+ *  \param[in]     stream          CUDA stream used for the operation.
+ */
+void nvte_fused_topk_with_score_function_forward(
+    const NVTETensor logits, int num_tokens, int num_experts, int topk, bool use_pre_softmax,
```
Comment (timmoon10) [inline transformer_engine/common/include/transformer_engine/fused_router.h]: Nit: The C API should not use `bool` since it is a C++ type:
```suggestion
    const NVTETensor logits, int num_tokens, int num_experts, int topk, int use_pre_softmax,
```

File: transformer_engine/common/fused_router/fused_moe_aux_loss.cu [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+}
+
+template <typename DataType, typename IndexType>
+void fused_moe_aux_loss_forward_kernel_launcher(const DataType* probs,
+                                                const IndexType* tokens_per_expert,
+                                                int total_num_tokens, int num_tokens,
+                                                int num_experts, int topk, float coeff,
+                                                DataType* aux_loss, float* Const_buf,
+                                                cudaStream_t stream) {
+  if (cuda::sm_arch(cuda::current_device()) >= 900) {
```
Comment (timmoon10) [inline transformer_engine/common/fused_router/fused_moe_aux_loss.cu:171]: Do we handle the CUDA stream properly in this path?
